### PR TITLE
fix: allow to disable stats color via --no-color

### LINKED ISSUE #2038: webpack-cli@v4.2.0 breaks color support on CI/CD
**Describe the bug**

Automated color detection was broken in v4.0.0 and subsequently fixed in v4.1.0. It is now broken again in v4.2.0 but only partially. The removal of `--color` flag also means we can not override this behavior and are left with uncolored output on CI.

**Expected behavior**

Either color support is properly detected or an override flag is provided again.

**Screenshots**

v4.1.0:
firefox_2020-11-04_21-44-42

v4.2.0:
firefox_2020-11-04_21-44-59

Please note that Drone fully supports ANSI colors but might not be properly detected by some tools (or outright blocked due to CI environment variable being detected)

### PR DESCRIPTION
**Pull request by snitin315**
**What kind of change does this PR introduce?**
bugfix

**Did you add tests for your changes?**
Yes
**If relevant, did you update the documentation?**
no
**Summary**

`stats` colors should be disabled if we use `--no-color`.

Fixes #2038  as well

### Before
Screenshot at 2020-11-05 09-11-53

### After
Screenshot at 2020-11-05 09-12-28

**Does this PR introduce a breaking change?**
Nope

**Other information**
No

### CODE REVIEW

**Comment by snitin315** [comment]
# from CI -

## webpack-4

Screenshot at 2020-11-05 09-56-31

Screenshot at 2020-11-05 09-57-00


## webpack-5
Screenshot at 2020-11-05 09-59-29

Screenshot at 2020-11-05 09-59-53

File: test/colors/colors.test.js [hunk 1]
```diff
@@ -43,6 +46,31 @@ describe('colorts', () => {
...
+        expect(stdout).toContain(output);
+        expect(exitCode).toBe(0);
+    });
+
+    it('should work with the "stats" option and --color flags', () => {
+        const { stderr, stdout, exitCode } = run(__dirname, ['--stats=verbose', '--color']);
+
+        //log to check in CI
+        console.log('colors should be enabled here -');
+        console.log(stdout);
```
Comment (evilebottnawi) [inline test/colors/colors.test.js:67]: You should avoid test colors like these, colors have special character, just check it on special characters

File: packages/webpack-cli/lib/webpack-cli.js [hunk 2]
```diff
@@ -235,6 +235,7 @@ class WebpackCLI {
                     }
                 }
 
+                coloretteOptions.enabled = args.color;
```
Comment (evilebottnawi) [inline packages/webpack-cli/lib/webpack-cli.js:238]: if (!coloretteOptions.enabled && args.color) { acoloretteOptions.enabled = args.color }

**Review by evilebottnawi — APPROVED** [review]
/cc @webpack/cli-team 

Good job, thanks

**Review by evilebottnawi — DISMISSED** [review]
oh, what about logger? Because `--no-colors` should disable colors for logger too

**Review by rishabh3112 — APPROVED** [review]

**Comment by snitin315** [comment]
> oh, what about logger? Because `--no-colors` should disable colors for logger too

I will update 👍🏻

**Comment by snitin315** [comment]
I will fix the CI problems soon.

File: packages/webpack-cli/lib/bootstrap.js [hunk 3]
```diff
@@ -23,6 +24,11 @@ const runCLI = async (cliArgs) => {
         // If the unknown arg starts with a '-', it will be considered an unknown flag rather than an entry
         let entry;
 
+        // enable/disable colors
+        if (coloretteOptions.enabled && !parsedArgs.opts.color) {
+            coloretteOptions.enabled = parsedArgs.opts.color;
+        }
```
Comment (evilebottnawi) [inline packages/webpack-cli/lib/bootstrap.js:30]: We should respect here `--colors` and `--no-colors`

Comment (snitin315) [reply packages/webpack-cli/lib/bootstrap.js:30]: I think we are or am I missing something?

Comment (evilebottnawi) [reply packages/webpack-cli/lib/bootstrap.js:30]: So if user provide the value we should respect it, if not colorette automatically enable/disable it
```js
if (typeof parsedArgs.opts.color !== 'undefined') {
  coloretteOptions.enabled = Boolean(parsedArgs.opts.color);
}
```

Comment (snitin315) [reply packages/webpack-cli/lib/bootstrap.js:30]: Actully we have `color: true` by default. So `parsedArgs.opts.color` is always true. 

https://github.com/webpack/webpack-cli/blob/46cba367f06a6354fe98fcb15e7771e819feeac0/packages/webpack-cli/lib/utils/cli-flags.js#L95

But we can add this check to make it more future proof.

Comment (evilebottnawi) [reply packages/webpack-cli/lib/bootstrap.js:30]: We should remove this default value, because it is invalid colors autodetecting inside `colorretter`

**Comment by evilebottnawi** [comment]
@snitin315 Can we add test for https://github.com/webpack/webpack-cli/issues/2054?

**Comment by snitin315** [comment]
Yes

**Review by evilebottnawi — APPROVED** [review]

**Review by evilebottnawi — DISMISSED** [review]
/cc @webpack/cli-team

File: packages/webpack-cli/lib/webpack-cli.js [hunk 4]
```diff
@@ -327,6 +327,11 @@ class WebpackCLI {
         let options = this.compilerConfiguration;
         let outputOptions = this.outputConfiguration;
 
+        // enable/disable colors
+        if (typeof args.color !== 'undefined') {
+            coloretteOptions.enabled = Boolean(args.color);
+        }
```
Comment (evilebottnawi) [inline packages/webpack-cli/lib/webpack-cli.js:333]: oh, stop, why we need it here? we already do it https://github.com/webpack/webpack-cli/pull/2042/files#diff-f7ca7db5e4f622982a83595e0f0bdef1a6704559c96f9b7602e1349fea350105R30

Comment (snitin315) [reply packages/webpack-cli/lib/webpack-cli.js:333]: Here its for stats.

Comment (evilebottnawi) [reply packages/webpack-cli/lib/webpack-cli.js:333]: Please clarify

Comment (snitin315) [reply packages/webpack-cli/lib/webpack-cli.js]: Leftover from an initial commit when I implemented it only for `stats`.

File: test/colors/colors.test.js [hunk 5]
```diff
@@ -43,6 +43,25 @@ describe('colorts', () => {
...
+    });
+
+    it('should work with the "stats" option and --color flags', () => {
+        const { stderr, stdout, exitCode } = run(__dirname, ['--stats=verbose', '--color']);
+
+        expect(stderr).toBeFalsy();
+        const output = isWebpack5 ? 'successfully' : 'main.js';
+        expect(stdout).toContain(`\u001b[1m\u001b[32m${output}\u001b[39m\u001b[22m`);
+        expect(exitCode).toBe(0);
+    });
```
Comment (evilebottnawi) [inline test/colors/colors.test.js:63]: Let's add couple test more:
- `stats: { colors: true }` and `--no-color`
- `stats: { colors: false }` and `--color`

**Review by evilebottnawi — COMMENTED** [review]
Good job, couple test and we can merge it

File: test/colors/colors.test.js (hunk 5 above)
Comment (snitin315) [reply test/colors/colors.test.js:63]: I think we prefer colors in config over --color. 
https://github.com/webpack/webpack-cli/blob/cd567127fe3651a9a7d0bc21cf79ae4a78c0e3c4/packages/webpack-cli/lib/webpack-cli.js#L370

Comment (anshumanv) [reply test/colors/colors.test.js:63]: nope flag should have preference

Comment (evilebottnawi) [reply test/colors/colors.test.js:63]: Yes, but let's add tests to avoid regressions

Comment (evilebottnawi) [reply test/colors/colors.test.js:63]: I think we should prefer `--color` over `stats.colors` in this case

Comment (evilebottnawi) [reply test/colors/colors.test.js:63]: Because I want to disable colors when run `--no-color`

Comment (evilebottnawi) [reply test/colors/colors.test.js:63]: And using `coloretteOptions.enabled` by default if no `stats.color` or `--color`/`--no-color`

Comment (snitin315) [reply test/colors/colors.test.js:63]:
```
stats.colors = typeof stats.colors !== 'undefined' &&  !args.colors ? stats.colors : coloretteOptions.enabled
```

Comment (evilebottnawi) [reply test/colors/colors.test.js:63]:
```js
let colors;

// From flags
if (typeof args.colors !== 'undefined') { 
  colors = args.colors;
} 
// From stats
else if (typeof stats.colors !== 'undefined') {
   colors = stats.colors;
} 
// Default
else {
 colors = coloretteOptions.enabled;
}
```

Comment (snitin315) [reply test/colors/colors.test.js:63]: done 💯

**Review by evilebottnawi — APPROVED** [review]
/cc @webpack/cli-team

**Comment by snitin315** [comment]
Good to merge?

**Review by evilebottnawi — APPROVED** [review]
/cc @webpack/cli-team

**Comment by anshumanv** [comment]
Thanks for waiting, I'm on it 👍

File: packages/webpack-cli/lib/bootstrap.js [hunk 6]
```diff
@@ -23,6 +24,11 @@ const runCLI = async (cliArgs) => {
         // If the unknown arg starts with a '-', it will be considered an unknown flag rather than an entry
         let entry;
 
+        // enable/disable colors
+        if (typeof parsedArgs.opts.color !== 'undefined') {
```
Comment (anshumanv) [inline packages/webpack-cli/lib/bootstrap.js:28]:
```suggestion
        if (parsedArgs.opts.hasOwnProperty("color")) {
```

File: packages/webpack-cli/lib/webpack-cli.js [hunk 7]
```diff
@@ -367,7 +367,21 @@ class WebpackCLI {
                     }
                 }
 
-                stats.colors = typeof stats.colors !== 'undefined' ? stats.colors : coloretteOptions.enabled;
+                let colors;
+                // From flags
+                if (typeof args.color !== 'undefined') {
```
Comment (anshumanv) [inline packages/webpack-cli/lib/webpack-cli.js:372]: update checks as above

**Review by anshumanv — CHANGES_REQUESTED** [review]
minor nits

File: packages/webpack-cli/lib/bootstrap.js (hunk 6 above)
Comment (snitin315) [reply packages/webpack-cli/lib/bootstrap.js:28]: Isn't it same ?

Comment (anshumanv) [reply packages/webpack-cli/lib/bootstrap.js:28]: nope, you can have a value `color: undefined`

Comment (snitin315) [reply packages/webpack-cli/lib/bootstrap.js:28]:
```js
 test = {
   colors: undefined
}

console.log(typeof test.colors === "undefined")
//output: true

console.log(test.hasOwnProperty("colors"))
//output true
```

Comment (snitin315) [reply packages/webpack-cli/lib/bootstrap.js:28]: Can you commit suggestion ? I'm off my system.

File: packages/webpack-cli/lib/webpack-cli.js (hunk 7 above)
Comment (snitin315) [reply packages/webpack-cli/lib/webpack-cli.js:372]:
```suggestion
                if (args.hasOwnProperty('color') {
```

File: packages/webpack-cli/lib/bootstrap.js [hunk 8]
```diff
@@ -23,6 +24,11 @@ const runCLI = async (cliArgs) => {
         // If the unknown arg starts with a '-', it will be considered an unknown flag rather than an entry
         let entry;
 
+        // enable/disable colors
+        if (parsedArgs.opts.hasOwnProperty('color')) {
```
Comment (evilebottnawi) [inline packages/webpack-cli/lib/bootstrap.js:28]: Please fix CI problem

Comment (evilebottnawi) [reply packages/webpack-cli/lib/bootstrap.js:28]: Why do you need `hasOwnProperty`? we don't use prototypes here, so just use `parsedArgs.opts.color`

Comment (snitin315) [reply packages/webpack-cli/lib/bootstrap.js:28]: https://github.com/webpack/webpack-cli/pull/2042#discussion_r519973255

Comment (snitin315) [reply packages/webpack-cli/lib/bootstrap.js:28]:
```suggestion
        if (typeof parsedArgs.opts.color !== 'undefined') {
```

File: packages/webpack-cli/lib/webpack-cli.js [hunk 9]
```diff
@@ -367,7 +367,21 @@ class WebpackCLI {
                     }
                 }
 
-                stats.colors = typeof stats.colors !== 'undefined' ? stats.colors : coloretteOptions.enabled;
+                let colors;
+                // From flags
+                if (args.hasOwnProperty('color')) {
```
Comment (snitin315) [inline packages/webpack-cli/lib/webpack-cli.js:372]:
```suggestion
                if (typeof args.color !== 'undefined') {
```

Comment (snitin315) [inline packages/webpack-cli/lib/webpack-cli.js:372]:
```suggestion
                if (typeof args.color !== 'undefined') {
```

**Review by anshumanv — APPROVED** [review]
Thanks

**Review by evilebottnawi — APPROVED** [review]
Feel free to merge after CI green

File: packages/webpack-cli/lib/bootstrap.js (hunk 8 above)
Comment (anshumanv) [reply packages/webpack-cli/lib/bootstrap.js]: strange that problems occur with hasOwnProperty, let's go with this for now

**Comment by evilebottnawi** [comment]
random fail on macos, let's ignore, we will return to this if it happens again

**Pull request merged**